### PR TITLE
use transform for execution to get torch_compile executable

### PR DESCRIPTION
**Pull request by t-vi**
Transform for execution used to produce duplicate names, this changes the implementation to be close to interpret_trace (with the plan of merging the two eventually).

Unfortunately, there seems to be some bad interaction, the newly skipped tests have been leading to segfaults, this needs investigation. 

Fixes: #1131
Also:
Fixes: #1501

### CODE REVIEW

File: thunder/core/trace_interpreter.py [hunk 1]
```diff
@@ -203,3 +204,167 @@ def do_swap(v):
...
+
+    def add_to_swap_map(self, old, new):
+        if old is new:
+            return
+        if isinstance(old, ProxyInterface):
+            if isinstance(new, ProxyInterface) and variableify(new) in self.env:
+                # the new isn't new, but something returned the input
+                # this means we need to map the old to the new
+                old, new = new, old
+            elif isinstance(old, TensorProxyInterface):
```
Comment (mruberry) [inline thunder/core/trace_interpreter.py:242]: Maybe in the future we won't need to have logic specific to VJPDuals, for example, here

File: thunder/core/transform_common.py [hunk 2]
```diff
@@ -111,6 +111,24 @@ def check(inp, log_str):
             check(copy_to_out, "output")
 
 
+def remove_duplicate_number_proxies(bsyms):
```
Comment (mruberry) [inline thunder/core/transform_common.py:114]: Comment or docstring here explaining the function would be nice

File: thunder/core/transform_common.py [hunk 3]
```diff
@@ -174,7 +192,9 @@ def _helper(x):
                 needed_proxies.add(variableify(x))
 
     dcetrace = from_trace(trace)
-    dcetrace.bound_symbols = list(reversed(dced))
+    dced_bound_symbols = list(reversed(dced))
+    dced_bound_symbols = dced_bound_symbols
```
Comment (mruberry) [inline thunder/core/transform_common.py:196]: What's up with this line?

File: thunder/core/transform_common.py [hunk 4]
```diff
@@ -174,7 +192,9 @@ def _helper(x):
...
+    dcetrace.bound_symbols = remove_duplicate_number_proxies(dced_bound_symbols)
```
Comment (mruberry) [inline thunder/core/transform_common.py:197]: This is so interesting. Would you add a comment here, or maybe to the description of the dce pass itself, about how duplicate number proxies might happen? What does it look like to have them and then to remove them?

fyi @jjsjann123

File: thunder/core/trace_interpreter.py [hunk 5]
```diff
@@ -203,3 +204,167 @@ def do_swap(v):
         return new_trace, tree_map(read, trace.output), env
 
     return new_trace, tree_map(read, trace.output)
+
+
+class TraceSubstitutionProcessor:
```
[the remark this answers is not in the document]
Comment (mruberry) [reply thunder/core/trace_interpreter.py:209]: (Maybe with a description of how it should be subclasses, like how the `OpExProcessor` subclasses it)

File: thunder/executors/passes.py [hunk 6]
```diff
@@ -61,68 +62,54 @@ def preserve_bsym(bsym: BoundSymbol) -> Any:
...
-    bound_symbols: list[BoundSymbol] = []
-    for bsym in extrace.bound_symbols:
-        nbsym: BoundSymbol = bsym.from_bsym_swap_proxies(swapmap)
-        bound_symbols.append(nbsym)
-
-    extrace.bound_symbols = bound_symbols
+
+    class OpExProcessor(TraceSubstitutionProcessor):
+        def process_bsym(self, bsym):
+            if bsym.sym.python_impl is not None:
```
Comment (mruberry) [inline thunder/executors/passes.py:68]: Comment for this short-circuit would be cool

File: thunder/executors/passes.py [hunk 7]
```diff
@@ -61,68 +62,54 @@ def preserve_bsym(bsym: BoundSymbol) -> Any:
...
+                        return
+                    else:
+                        raise AssertionError("Unknown executor")
+
+            if bsym.sym.executor is not None:
+                self.add_processed_bsyms([bsym])
+                self.set_result(bsym.output)
+                return
+
+            # No executor found, need to descend
```
Comment (mruberry) [inline thunder/executors/passes.py:107]: Nice comment

File: thunder/executors/torch_autograd.py [hunk 8]
```diff
@@ -240,6 +240,23 @@ def split_forward_backward(computation_trc: TraceCtx, compile_data, compile_stat
         skip_output=False,
         skip_subsymbols=False,
     )
+
+    # remove duplicates
+    # The NVFuser (and possibly others) fusion pass applied on the forward during has a
```
Comment (mruberry) [inline thunder/executors/torch_autograd.py:245]: This is super interesting!

fyi @jjsjann123 @kshitij12345 @IvanYashchuk

File: thunder/executors/torch_compile.py [hunk 9]
```diff
@@ -63,36 +68,34 @@ def make_compiled(
...
+        for a in sorted_unique_inputs:
+            prims.unpack_trivial(a, name=a.name)
+
+    region_trace.bound_symbols += list(bsyms)
     region_trace.bound_symbols.append(prims.python_return.bind(sorted_unique_outputs, output=None))
+    # for a in region_trace.args:
+    #    region_trace.add_name(a.name)
+    for bsym in region_trace.bound_symbols:
+        for o in bsym.flat_outs:
+            if o is not None:  # TODO: investigate
```
Comment (mruberry) [inline thunder/executors/torch_compile.py:89]: Would you elaborate on this TODO -- is the question about what's happening with a bsym with no outputs?

File: thunder/torch/__init__.py [hunk 10]
```diff
@@ -5754,6 +5754,7 @@ def register_default_torch_op(torchfn: Callable, torch_module):
         name=torchfn_name,
         meta=_fn,
         id=f"{torch_module.__name__}.{torchfn_name}",
+        is_prim=True,
```
Comment (mruberry) [inline thunder/torch/__init__.py:5757]: Interesting! Why do we want to change this? Won't this default torch operations to be primitives? What's the effect of that?

File: thunder/core/trace_interpreter.py (hunk 1 above)
Comment (t-vi) [reply thunder/core/trace_interpreter.py:242]: The irony is that `VJPDuals` neither have VJP-specific logic nor is there any duality involved, so this should probably just be called `AugmentedValue`, I will eventually send a PR to rename it.
As the next thing is to consolidate `interpret_trace_to_trace` into this, let's keep it for now.

File: thunder/core/transform_common.py (hunk 3 above)
Comment (t-vi) [reply thunder/core/transform_common.py:196]: Ahahah. I prepared the next line and then forgot the preparations.

File: thunder/core/transform_common.py (hunk 4 above)
Comment (t-vi) [reply thunder/core/transform_common.py:197]: Well, so basically remove the line and run the tests.

File: thunder/executors/passes.py (hunk 6 above)
Comment (t-vi) [reply thunder/executors/passes.py:68]: Yeah, I copy-pasted this from your version of the visitor logic. What would be the comment?

File: thunder/torch/__init__.py (hunk 10 above)
Comment (t-vi) [reply thunder/torch/__init__.py:5757]: Note that these are the auto-registered operations that don't have a decomposition into subsymbols. Arguably, this has been a bug before, as dce on a symbol with auto-registered operations as subsymbols would remove them.

File: thunder/executors/torch_compile.py (hunk 9 above)
Comment (t-vi) [reply thunder/executors/torch_compile.py:89]: I think the todo can be dropped. I was surprised that we get None outputs here, but it is OK.

File: thunder/executors/passes.py [hunk 11]
```diff
@@ -61,68 +62,54 @@ def preserve_bsym(bsym: BoundSymbol) -> Any:
...
-    # Restores original variables
-    bound_symbols: list[BoundSymbol] = []
-    for bsym in extrace.bound_symbols:
-        nbsym: BoundSymbol = bsym.from_bsym_swap_proxies(swapmap)
-        bound_symbols.append(nbsym)
-
-    extrace.bound_symbols = bound_symbols
+
+    class OpExProcessor(TraceSubstitutionProcessor):
+        def process_bsym(self, bsym):
```
[the remark this answers is not in the document]
Comment (t-vi) [reply thunder/executors/passes.py:61]: Added a comment.

File: thunder/core/trace_interpreter.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (t-vi) [reply thunder/core/trace_interpreter.py:209]: Added a docstring with a reference to the OpExProcessor.

File: thunder/core/trace_interpreter.py [hunk 12]
```diff
@@ -203,3 +204,179 @@ def do_swap(v):
...
+        if isinstance(v, VJPDual):
+            v.primal = tree_map(self.do_swap, v.primal)
+            v.residuals = tree_map(self.do_swap, v.residuals)
+            return v
+        if not isinstance(v, ProxyInterface):
+            return v
+        return self.swap_map.get(variableify(v), v)
+
+    def add_unprocessed_bsyms(self, bsyms):
+        self.unprocessed_bsyms[:0] = bsyms
```
Comment (lantiga) [inline thunder/core/trace_interpreter.py:284]: we could probably add `self.unprocessed_bsyms` and `self.new_bsyms` to `__init__` for readability - not mandatory

**Review by lantiga — APPROVED** [review]
Looks great, also great docstring addition! Let's merge as is

File: thunder/executors/torch_autograd.py (hunk 8 above)
Comment (jjsjann123) [reply thunder/executors/torch_autograd.py:245]: Out of curiosity, do duplicates cause any functional issue?

**Pull request merged**